### PR TITLE
Feature multilayer perceptron and data-driven fluid model

### PR DESCRIPTION
**Pull request by EvertBunschoten**
## Proposed Changes
Addition of multi-layer perceptron class which can be used to evaluate trained multi-layer perceptrons in processes such as thermodynamic state evaluation in data-driven fluid models.
 

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

In the Tutorials repository, branch feature_multilayer_perceptron, one can find under "common/multilayer_perceptron" some python scripts which demonstrate how to translate a model trained through Tensorflow to an input file compatible with the CLookUp_ANN class. The README.md file in that tutorial folder also lists some important information regarding the functionality of the CLookUp_ANN class, datadriven fluid model, and MLP input file format. 

Shall I leave those there or should these be uploaded elsewhere?

- [X] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [x] I have added a test case that demonstrates my contribution, if necessary.
- [x] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Comment by pcarruscag** [comment]
Do you want to introduce this first as a standalone library, and then start using it for fluid models?

**Comment by EvertBunschoten** [comment]
@pcarruscag Yes I want to introduce the multilayer perceptron as an option for users and developers for things like data-driven fluid models. Do you think it would be good to add a template CFluidModel child class demonstrating how the MLP class can be used to create data-driven fluid models (apart from writing a tutorial of course)?

**Comment by pcarruscag** [comment]
Sounds good. Initially it would be enough to have some unit tests that would already show how to setup the network, together with documentation / example of the file format. Applications can come after.

Some initial comments:
Please move the files to toolboxes/ and ideally use a namespace for the new classes.
Start the class names with C as we do, e.g. CIOMap.

Then, how large are the models you've used so far? and how important is performance to this feature? (Just so I know how much to comment on that)

**Comment by EvertBunschoten** [comment]
Of course I'll provide an example of an MLP input file, as well as a python script I wrote that translates an MLP trained through TensorFlow to such an input file. 

Very well, I'll move the files from the numerics folder to the common folder and change names accordingly.

The models I used so far had between 5 and 50 perceptrons and up to 15 layers. Performance is quite important, as evaluation of MLP's is generally more computationally expensive than for example lookup tables. The larger the MLP architecture, the more costly evaluations get of course. Any improvement to the computation speed will therefore be welcome. In terms of memory, the MLP's don't seem to be an issue so far.

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include <string>
+#include <iostream>
+#include <limits>
+#include <cstdlib>
+#include <vector>
+#include "../../CConfig.hpp"
+#include "../../linear_algebra/blas_structure.hpp"
+#include "CLookUp_ANN.hpp"
+
+using namespace std;
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:39]: Dont relly on this please, just type std:: where needed

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+#pragma once
+
+#include <cmath>
+#include <string>
+#include <iostream>
+#include <limits>
+#include <cstdlib>
+#include <vector>
+#include "../../CConfig.hpp"
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:35]: Why is the config needed here?

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#include "../../linear_algebra/blas_structure.hpp"
+#include "CLookUp_ANN.hpp"
+
+using namespace std;
+namespace MLPToolbox
+{
+        
+    class CLookUp_ANN;
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:43]: Why do have a forward declaration if you include the corresponding header?

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    class CIOMap
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:45]: Please run clang-format on new files and add documentation at least at the class level, i.e. what is the job of this class? Its a map of what into what?

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        private:
+        vector<size_t> ANN_indices;
+        vector<vector<pair<size_t, size_t>>> Input_Map;
+        vector<vector<pair<size_t, size_t>>> Output_Map;
+        vector<vector<su2double>> input_indices;
+        public:
+        CIOMap(){};
+        CIOMap(CLookUp_ANN*ANN_collection, vector<string> &inputs, vector<string> &outputs);
+        ~CIOMap(){};
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:55]: Please have a look at the c++ guidelines regarding contructors and destructors.

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        vector<pair<size_t, size_t>> GetOutputMapping(size_t i_map){return Output_Map[i_map];}
+        vector<pair<size_t, size_t>> GetInputMapping(size_t i_map){return Input_Map[i_map];}
+        
+        vector<su2double> GetANN_Inputs(size_t i_Map, vector<su2double>&inputs){
+            vector<su2double> ANN_input;
+            ANN_input.resize(Input_Map[i_Map].size());
+
+            for(size_t iInput=0; iInput<Input_Map[i_Map].size(); iInput++){
+                ANN_input.at(iInput) = inputs[GetInputIndex(i_Map, iInput)];
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:73]: Dont use 'at' for std::vector

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        vector<vector<pair<size_t, size_t>>> Output_Map;
+        vector<vector<su2double>> input_indices;
+        public:
+        CIOMap(){};
+        CIOMap(CLookUp_ANN*ANN_collection, vector<string> &inputs, vector<string> &outputs);
+        ~CIOMap(){};
+        void Push_ANN_index(size_t i_ANN){ANN_indices.push_back(i_ANN);};
+        void PairVariableswithANNs(CLookUp_ANN * ANN_collection, vector<string> &inputs, vector<string> &outputs);
+        size_t GetNANNs(){return ANN_indices.size();}
+        size_t GetANNIndex(size_t i_Map){return ANN_indices[i_Map];}
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:59]: Please have a look at the c++ guidelines for const correctness

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        size_t GetInputIndex(size_t i_Map, size_t iInput){return Input_Map[i_Map][iInput].first;}
+        size_t GetOutputIndex(size_t i_Map, size_t iOutput){return Output_Map[i_Map][iOutput].first;}
+        size_t GetANNOutputIndex(size_t i_Map, size_t iOutput){return Output_Map[i_Map][iOutput].second;}
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:62]: It would be nice to pick a consistent style for your new files, please consider doing some replace alls. I tried to list some options that are consistent with parts of SU2 at least.

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp (hunk 3 above)
Comment (EvertBunschoten) [reply Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:43]: The CIOMap class uses a pointer to a CLookUp_ANN object in the constructor. Somehow when I remove the forward declaration, I'm getting a compiler error.

File: Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+#include <cmath>
+#include <string>
+#include <iostream>
+#include <limits>
+#include <cstdlib>
+#include <vector>
+#include "../../CConfig.hpp"
+#include "../../linear_algebra/blas_structure.hpp"
+#include "CNeuralNetwork.hpp"
+#include "CIOMap.hpp"
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp:39]: Well you have a cyclic dependency between the two classes

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+#include "CLookUp_ANN.hpp"
+
+namespace MLPToolbox {
+
+class CLookUp_ANN;
+
+class CIOMap
+/*!
+ *\class CIOMap
+ *\brief This class is used by the CLookUp_ANN class to assign user-defined inputs
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:46]: Documentation above declaration, please add a doxygen group in docmain.hpp and place these classes in that group.

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+  void Push_MLP_index(std::size_t iMLP) { MLP_indices.push_back(iMLP); }
+
+  /*!
+  * \brief Pair call inputs and outputs with the inputs and outputs of
+          the loaded MLPs
+  * \param[in] MLP_collection - pointer to MLP collection class
+  * \param[in] inputs - vector with call input variable names
+  * \param[in] outputs - vector with call output variable names
+  */
+  void PairVariableswithMLPs(CLookUp_ANN* MLP_collection, su2vector<std::string>& inputs,
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:110]: Please remove the underscores in the middle of CamelCase names. It is not consistent within this class.

File: Common/include/toolboxes/multilayer_perceptron/CLayer.hpp [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+#include "../../CConfig.hpp"
+#include "../../linear_algebra/blas_structure.hpp"
+#include "CNeuron.hpp"
+
+namespace MLPToolbox {
+class CLayer {
+ private:
+  unsigned long number_of_neurons; /*!< Neuron count in current layer */
+  CNeuron* neurons;                /*!< Array of neurons in current layer */
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLayer.hpp:42]: Use a vector

File: Common/include/toolboxes/multilayer_perceptron/CLayer.hpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+
+#include "../../CConfig.hpp"
+#include "../../linear_algebra/blas_structure.hpp"
+#include "CNeuron.hpp"
+
+namespace MLPToolbox {
+class CLayer {
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLayer.hpp:39]: Documentation

File: Common/include/toolboxes/multilayer_perceptron/CLayer.hpp [hunk 14]
```diff
@@ -0,0 +1,146 @@
+/*!
+ * \file CLayer.hpp
+ * \brief Declaration of artificial neural network interpolation class
+ * \author E. Bunschoten
+ * \version 7.5.0 "Blackbird"
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLayer.hpp:5]: In other files too please.
```suggestion
 * \version 7.5.1 "Blackbird"
```

File: Common/include/toolboxes/multilayer_perceptron/CLayer.hpp [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2022, SU2 Contributors (cf. AUTHORS.md)
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLayer.hpp:12]:
```suggestion
 * Copyright 2012-2023, SU2 Contributors (cf. AUTHORS.md)
```

File: Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+   * thermodynamic state look-up operations. The multi-layer perceptrons are loaded in the order listed
+   * in the MLP collection file. Each multi-layer perceptron is generated based on the
+   * architecture described in its respective input file. When evaluating the MLP collection, an
+   * input-output map is used to find the correct MLP corresponding to the call function inputs and
+   * outputs.
+   */
+
+ private:
+  int rank{0};
+  su2vector<CNeuralNetwork*> NeuralNetworks; /*!< std::vector containing all loaded neural networks. */
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp:57]: Why is this done with pointers? Can you store the objects directly?

File: Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+   * \param[in] input_output_map - pointer to input-output map to be checked
+   */
+  bool Check_Use_of_Outputs(su2vector<std::string>& output_names, CIOMap* input_output_map) const;
+
+  /*!
+   * \brief Check if all input variables are present in the loaded ANNs
+   * \param[in] input_names - input variable names to check
+   * \param[in] input_output_map - pointer to input-output map to be checked
+   */
+  bool Check_Use_of_Inputs(su2vector<std::string>& input_names, CIOMap* input_output_map) const;
```
Comment (pcarruscag) [inline Common/include/toolboxes/multilayer_perceptron/CLookUp_ANN.hpp:107]: more underscores

File: Common/src/toolboxes/multilayer_perceptron/CNeuralNetwork.cpp [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+  if (input.compare("linear") == 0) {
+    activation_function_types[i_layer] = ENUM_ACTIVATION_FUNCTION::LINEAR;
+    return;
+  }
+  if (input.compare("relu") == 0) {
+    activation_function_types[i_layer] = ENUM_ACTIVATION_FUNCTION::RELU;
+    return;
+  }
+  if ((input.compare("elu") == 0)) {
+    activation_function_types[i_layer] = ENUM_ACTIVATION_FUNCTION::ELU;
```
Comment (pcarruscag) [inline Common/src/toolboxes/multilayer_perceptron/CNeuralNetwork.cpp:367]: use a map

File: Common/src/toolboxes/multilayer_perceptron/CNeuron.cpp [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "../../../include/toolboxes/multilayer_perceptron/CNeuron.hpp"
+
+#include <cmath>
+#include <cstring>
+#include <iostream>
+
+using namespace std;
```
Comment (pcarruscag) [inline Common/src/toolboxes/multilayer_perceptron/CNeuron.cpp:34]: Why is this file needed?

**Review by pcarruscag — COMMENTED** [review]
It looks like the MLP is fairly independent of SU2 (except from MPI_Error ?) Have you considered including this as a subproject? By hosting the MLP as a personal repo?

File: Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp (hunk 10 above)
Comment (EvertBunschoten) [reply Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:46]: I made the MLP classes indeed as general as possible, such that they could be used in a variety of applications within SU2 (fluid models, sub-grid scale models, etc). Apart from MPI, variable types like su2double, MapType, and su2vector are used. I have not considered hosing the MLP as a personal repo. What would be the benefit of this?

Comment (EvertBunschoten) [reply Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:46]: > Documentation above declaration, please add a doxygen group in docmain.hpp and place these classes in that group.

What do you mean exactly by "documentation above declaration"?

Comment (pcarruscag) [reply Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:46]: "Documentation"
class Class {
...

Comment (pcarruscag) [reply Common/include/toolboxes/multilayer_perceptron/CIOMap.hpp:46]: Less non-CFD code for us to review.
Easier for you or someone else to re-use in other projects.
The types are easy to solve if you make it header-only, the user can typedef before including the main header. Or you can template the classes.

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 20]
```diff
@@ -6502,6 +6507,9 @@ void CEulerSolver::BC_Giles(CGeometry *geometry, CSolver **solver_container, CNu
           Jacobian.SubtractBlock2Diag(iPoint, residual.jacobian_i);
 
       }
+      /*--- Store number of Newton iterations at BC ---*/
+      if(config->GetKind_FluidModel() == DATADRIVEN_FLUID)
+        GetNodes()->SetNewtonSolverIterations(iPoint, GetFluidModel()->GetnIter_Newton());
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:6512]: Inside the solvers you use "nodes" directly
```suggestion
        nodes->SetNewtonSolverIterations(iPoint, GetFluidModel()->GetnIter_Newton());
```

File: SU2_CFD/include/fluid/CFluidModel.hpp [hunk 21]
```diff
@@ -325,4 +325,28 @@ class CFluidModel {
...
+  * \brief Set the initial guess for the density in Newton solvers
+  * \param[in] rho - Initial value for density.
+  */
+  virtual void SetDensity(su2double rho) {}
+
+  /*!
+  * \brief Set the initial guess for the static energy in Newton solvers
+  * \param[in] e - Initial value for static energy.
+  */
+  virtual void SetEnergy(su2double e) {}
```
Comment (pcarruscag) [inline SU2_CFD/include/fluid/CFluidModel.hpp:339]: Put initial in the name, it is not clear what this does otherwise

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 22]
```diff
@@ -6125,6 +6150,11 @@ void CEulerSolver::BC_Giles(CGeometry *geometry, CSolver **solver_container, CNu
       Energy_i = nodes->GetEnergy(iPoint);
       StaticEnergy_i = Energy_i - 0.5*Velocity2_i;
 
+      /*--- Set initial values for density and energy for Newton solvers in fluid model ---*/
+      su2double relax_Newton = config->GetRelaxation_DataDriven();
+      GetFluidModel()->SetInitialDensity(relax_Newton*Density_i + (1 - relax_Newton)*config->GetDensity_Init_DataDriven());
+      GetFluidModel()->SetInitialEnergy(relax_Newton*StaticEnergy_i + (1 - relax_Newton)*config->GetEnergy_Init_DataDriven());
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:6156]: `if(config->GetKind_FluidModel() == DATADRIVEN_FLUID)`

File: SU2_CFD/src/solvers/CEulerSolver.cpp [hunk 23]
```diff
@@ -4668,6 +4681,13 @@ void CEulerSolver::BC_Riemann(CGeometry *geometry, CSolver **solver_container,
 
       GetFluidModel()->SetTDState_rhoe(Density_i, StaticEnergy_i);
 
+      /*--- Set initial values for density and energy for Newton solvers in fluid model ---*/
+      if(config->GetKind_FluidModel() == DATADRIVEN_FLUID){
+        su2double relax_Newton = config->GetRelaxation_DataDriven();
+        GetFluidModel()->SetInitialDensity(relax_Newton*Density_i + (1 - relax_Newton)*config->GetDensity_Init_DataDriven());
+        GetFluidModel()->SetInitialEnergy(relax_Newton*StaticEnergy_i + (1 - relax_Newton)*config->GetEnergy_Init_DataDriven());
+      }
+
```
Comment (pcarruscag) [inline SU2_CFD/src/solvers/CEulerSolver.cpp:4690]: Why is this used for boundary conditions but not when computing primitives for the entire domain?

**Review by pcarruscag — COMMENTED** [review]
Thank you for doing this as submodule, you need to use a compatible license for your code. The way you are handling the type is not safe, if we make changes the compilation will break, use one of the two ways I replied to you on Slack

**Review by pcarruscag — COMMENTED** [review]
Just in case you didn't notice you have the wrong versions of almost all of the submodules.

**Comment by EvertBunschoten** [comment]
> Thank you for doing this as submodule, you need to use a compatible license for your code. The way you are handling the type is not safe, if we make changes the compilation will break, use one of the two ways I replied to you

**Comment by EvertBunschoten** [comment]
> 

Hi Pedro, thanks for your reply. I don't know how to safely typedef the mlpdouble type the way you suggested. I thought you meant that I just had to copy the way su2double was defined in `code_config.hpp`, but I guess that's not right. Do you know an example showing how this works?

**Comment by pcarruscag** [comment]
in your type config file do:
```
#ifdef MLP_CUSTOM_TYPE
using mlpdouble = MLP_CUSTOM_TYPE;
#else
using mlpdouble = double;
#endif
```
in SU2 before the first MLP include do:
```
#define MLP_CUSTOM_TYPE su2double
#include ...
```

**Comment by EvertBunschoten** [comment]
> 

Thanks Pedro!

File: Common/include/CConfig.hpp [hunk 24]
```diff
@@ -3768,6 +3776,40 @@ class CConfig {
...
+  /*!
+   * \brief Datadriven method for EoS evaluation.
+   */
+  ENUM_DATADRIVEN_METHOD GetKind_DataDriven_Method(void) const { return Kind_DataDriven_Method; }
+
+  /*!
+   * \brief Get name of the input file for the data-driven fluid model interpolation method.
+   * \return Name of the input file for the interpolation method.
+   */
+  string* GetDataDriven_FileNames(void) const { return DataDriven_Method_FileNames; }
```
Comment (pcarruscag) [inline Common/include/CConfig.hpp:3788]:
```suggestion
  const string* GetDataDriven_FileNames(void) const { return DataDriven_Method_FileNames; }
```

File: SU2_CFD/src/solvers/CEulerSolver.cpp (hunk 23 above)
Comment (pcarruscag) [reply SU2_CFD/src/solvers/CEulerSolver.cpp:4687]: Unresolving again.

Comment (pcarruscag) [reply SU2_CFD/src/solvers/CEulerSolver.cpp]: @EvertBunschoten if this makes it to 2 months without a reply I'll start reverting what was merged into TestCases already.

**Comment by bigfooted** [comment]
@EvertBunschoten I suggest to first finish and close this one before moving on to the hydrogen flamelets PR

**Review by pcarruscag — CHANGES_REQUESTED** [review]
I'll ignore the request for review the same way you ignored my questions :shrug:

**Comment by EvertBunschoten** [comment]
> I'll ignore the request for review the same way you ignored my questions shrug

Could you clarify your questions? I sent you a message on Slack.

**Review by pcarruscag — APPROVED** [review]
Evert thought he was replying but the messages were pending and never made it online.
